// @sanity/client/dist/stega.browser.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var browserMiddleware = require('./_chunks/browserMiddleware-IWUpjfF7.cjs');
var getIt = require('get-it');
var stegaEncodeSourceMap = require('./_chunks/stegaEncodeSourceMap-YNx-0kzT.cjs');

class SanityStegaClient extends browserMiddleware.SanityClient {
}
class ObservableSanityStegaClient extends browserMiddleware.ObservableSanityClient {
}

const exp = browserMiddleware.defineCreateClientExports(browserMiddleware.envMiddleware, browserMiddleware.SanityClient);
const requester = exp.requester;
const createClient = exp.createClient;

exports.BasePatch = browserMiddleware.BasePatch;
exports.BaseTransaction = browserMiddleware.BaseTransaction;
exports.ClientError = browserMiddleware.ClientError;
exports.ObservablePatch = browserMiddleware.ObservablePatch;
exports.ObservableSanityClient = browserMiddleware.ObservableSanityClient;
exports.ObservableTransaction = browserMiddleware.ObservableTransaction;
exports.Patch = browserMiddleware.Patch;
exports.SanityClient = browserMiddleware.SanityClient;
exports.ServerError = browserMiddleware.ServerError;
exports.Transaction = browserMiddleware.Transaction;
exports.vercelStegaCleanAll = browserMiddleware.vercelStegaCleanAll;
Object.defineProperty(exports, "unstable__adapter", {
	enumerable: true,
	get: function () { return getIt.adapter; }
});
Object.defineProperty(exports, "unstable__environment", {
	enumerable: true,
	get: function () { return getIt.environment; }
});
exports.encodeIntoResult = stegaEncodeSourceMap.encodeIntoResult;
exports.stegaEncodeSourceMap = stegaEncodeSourceMap.stegaEncodeSourceMap;
exports.ObservableSanityStegaClient = ObservableSanityStegaClient;
exports.SanityStegaClient = SanityStegaClient;
exports.createClient = createClient;
exports.requester = requester;
