// @sanity/client/dist/stega.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var nodeMiddleware = require('./_chunks/nodeMiddleware-Gh18jsRe.cjs');
var getIt = require('get-it');
var stegaEncodeSourceMap = require('./_chunks/stegaEncodeSourceMap-ZDmZYDbM.cjs');

class SanityStegaClient extends nodeMiddleware.SanityClient {
}
class ObservableSanityStegaClient extends nodeMiddleware.ObservableSanityClient {
}

const exp = nodeMiddleware.defineCreateClientExports(nodeMiddleware.middleware, nodeMiddleware.SanityClient);
const requester = exp.requester;
const createClient = exp.createClient;

exports.BasePatch = nodeMiddleware.BasePatch;
exports.BaseTransaction = nodeMiddleware.BaseTransaction;
exports.ClientError = nodeMiddleware.ClientError;
exports.ObservablePatch = nodeMiddleware.ObservablePatch;
exports.ObservableSanityClient = nodeMiddleware.ObservableSanityClient;
exports.ObservableTransaction = nodeMiddleware.ObservableTransaction;
exports.Patch = nodeMiddleware.Patch;
exports.SanityClient = nodeMiddleware.SanityClient;
exports.ServerError = nodeMiddleware.ServerError;
exports.Transaction = nodeMiddleware.Transaction;
exports.vercelStegaCleanAll = nodeMiddleware.vercelStegaCleanAll;
Object.defineProperty(exports, "unstable__adapter", {
	enumerable: true,
	get: function () { return getIt.adapter; }
});
Object.defineProperty(exports, "unstable__environment", {
	enumerable: true,
	get: function () { return getIt.environment; }
});
exports.encodeIntoResult = stegaEncodeSourceMap.encodeIntoResult;
exports.stegaEncodeSourceMap = stegaEncodeSourceMap.stegaEncodeSourceMap;
exports.ObservableSanityStegaClient = ObservableSanityStegaClient;
exports.SanityStegaClient = SanityStegaClient;
exports.createClient = createClient;
exports.requester = requester;
